Zigbee improvements

oleo65 opened the pull request:
* Fixed battery cluster reporting (bug).
* Made some configuration properties configurable through Kconfig.
* Made log level configurable in zigbee and core parasite lib.
* Split debug build config from production.

I contributed to a zigbee firmware fork in the past and tried to migrate some concepts / bugfixes I experienced there. I personally really like the idea here, that the core libraries are reused no matter what protocol stack is used. This seems to me the right approach and I am happy to push the zigbee firmware forward.

In the long run, I believe even more configuration could be centralized. The Zephyr configuration system is quite powerful. Let me know what you think. 😃

oleo65 committed:
Making core log level configurable via Kconfig entry.

oleo65 committed:
Making zigbee main log level configurable.

oleo65 committed:
Fixing sensor logging millivolts.

oleo65 committed:
Fixed bug in cluster list not connecting power cluster.

oleo65 committed:
Improved basic attribute list to include more properties.

oleo65 committed:
Made more basic attributes compile time configurable.

oleo65 committed:
Added separate debug config to better separate flags during development.

rbaron reviewed (commented):
Hey @oleo65 that's fantastic. Thank you! I was looking forward to discuss this PR with someone. I've made some comments mostly about naming. Looking forward to your thoughts.

BTW out of curiosity, did you manage to test this with either Zigbee2MQTT or something else?

Hedda commented:
> I personally really like the idea here, that the core libraries are reused no matter what protocol stack is used. This seems to me the right approach and I am happy to push the zigbee firmware forward.

+1 The same core application for Zigbee should in theory be reusable for  Matter standard over Thread (Project CHIP), see -> https://github.com/rbaron/b-parasite/issues/87

oleo65 commented:
@rbaron I set the sensor up via nativ Home Assistant ZHA integration. as all my other devices. It basically worked straight out of the box.

If you already run a Zigbee2MQTT gateway, that might also be a valid solution, but I like to keep my setup simpler without relying on an additional (MQTT) component. I got away with it so far... 😉 

As for the naming of variables etc. I am quite torn in half. To me it felt almost all the time redundant to name something with `PRST` prefix or something similar. The code is for a firmware for the parasite and I believe it is not necessary to distinct "your own custom" code from "framework" code. But that is also a highly debatable topic on how to name your stuff in your code. (We often have that discussion at work, but usually resolve it with clear and simple names...)

I am fine to rename any variables in this pull request so that it complies with the rest of the project to move this forward. But I would love to start a general refactoring branch with more separations of concerns within the different code files to make it easier for new people (including me) to comprehend which code part does what. 😃

oleo65 commented on code/nrf-connect/samples/zigbee/Kconfig:
```diff
@@ -6,4 +6,12 @@ config PRST_ZB_SLEEP_DURATION_SEC
+  string "Zigbee app init basic date code. Can be used as build revision."
+  default "20221214"
+
+config MODEL_ID
```
I would advise against this as the model id is displayed quite prominently at least in HA. I see it more as the general model (piece of hardware) which to me is `b-parasite`. For the revision one could use a different property in the basic cluster maybe `hw_version`?

rbaron replied:
That makes sense, `hw_version` sounds like a much better place for it.

oleo65 replied:
Added config flag for `PRST_ZB_HARDWARE_VERSION`. This is only a `uint8` value and I set it per default to `2` as in `1.2.0`.

oleo65 commented on code/nrf-connect/samples/zigbee/prj.conf:
```diff
@@ -43,4 +43,6 @@ CONFIG_ZIGBEE_CHANNEL_SELECTION_MODE_MULTI=y
 
 # Config options in Kconfig.
-# CONFIG_PRST_ZB_SLEEP_DURATION_SEC=5
+CONFIG_PRST_ZB_SLEEP_DURATION_SEC=600
```
I would go even further as I understand the Kconfig pattern.

The sane default is set in the Kconfig file (as 600) and the default `prj.conf` does not contain it at all, so the default is used. If you decide to change it, you can by simply adding it to your `prf.conf` but the file stays cleaner without any commented possible options. 😃

oleo65 replied:
I liked the idea to name the variant `z-parasite` but you are right, that it is nowhere mentioned. To me it is still helpful as I have many different variants deployed and like the possibility to distinguish them. But I revert the default value for the project.

oleo65 commented on code/nrf-connect/samples/zigbee/Kconfig:
```diff
@@ -6,4 +6,12 @@ config PRST_ZB_SLEEP_DURATION_SEC
\ No newline at end of file
+	default 60
+
+config BUILD_DATE
```
As I understand the `BUILD_DATE` property, it is expected a 1-byte string and suggests to have a date stamp. 

One could either provide a fixed default date in code for people who don't care about it or maybe set it to the date of fixed github releases each time a new firmware is released?

In my local workflow I usually override it so that I can distinguish the firmware build on the deployed parasites...

rbaron replied:
That makes sense. I will make sure to add `-DPRST_ZB_BUILD_DATE` to the github action that builds this sample too.

oleo65 replied:
Changed the naming of the config flags as discussed.

oleo65 commented on code/nrf-connect/prstlib/boards/arm/bparasite_nrf52840/Kconfig:
```diff
@@ -11,4 +11,8 @@ config BOARD_ENABLE_DCDC_HV
\ No newline at end of file
+endif  # BOARD_BPARASITE_NRF52840
+
+config CORE_LOG_LEVEL
```
I don't know about the enum values but would guess not because the Kconfig is resolved first before any C code is compiled. But I did not research this further...

I thought about the config hierarchy and did not come up with a nice solution without duplicating code, since there is no real common ancestor in the device tree. I don't know if it is maybe possible to have more than one ancestor (like a mixin) and maybe have a common `b-parasite` mixin with common configuration flags for both MCU variants. I would really like that as it might help to reduce configuration errors and feature drifts...

rbaron replied:
I did a quick test:

```
config PRSTLIB_DEFAULT_LOG_LEVEL
	int "Default log level for prstlib"
	default LOG_LEVEL_INF
```
And unfortunately that doesn't work :( (`LOG_LEVEL_INF` is undefined at this point). It seems like Zephyr implements the following strategy for subsystem logging:

<img width="539" alt="image" src="https://user-images.githubusercontent.com/1573409/208300481-5a312b0c-7ea2-4340-aa88-3abb64ae7ae2.png">

Each subsystem defines their own choices. I'm not sure if this is overkill for us, but it would perfect if we could just add a `Kconfig` file to `prstlib` and do it there.

oleo65 replied:
Changed the naming of the config flag as discussed.

oleo65 commented:
I am pretty sure that I will eventually move to Matter over Thread. And I am happy that all the improvements to zigbee and the move to zephyr pay directly towards a Matter/Thread implementation.

At the moment I simply do not have the infrastructure to test any of this myself...

rbaron commented:
> As for the naming of variables etc. I am quite torn in half. To me it felt almost all the time redundant to name something with `PRST` prefix or something similar. The code is for a firmware for the parasite and I believe it is not necessary to distinct "your own custom" code from "framework" code. But that is also a highly debatable topic on how to name your stuff in your code. (We often have that discussion at work, but usually resolve it with clear and simple names...)

Agreed, naming is subjective and personal and I also like it as simple and short as possible, but no simpler. I think even in our simple setup there's already room for confusion: does `CORE_LOG_LEVEL` control the log level of the `prstlib` or the application or some Zephyr subsystems? Or all of them? If some other app includes `prstlib` as one of its few external libraries, does the name `CORE` still make sense? I would think it's clearer to have separate `PRSTLIB_DEFAULT_LOG_LEVEL` for the lib and `PRST_ZB_DEFAULT_LOG_LEVEL` for the application, and it's easier to grep them and see where they're modified.

oleo65 committed:
Changed name of config value for prstlib default log level.

oleo65 committed:
Fixed default sleep duration in Kconfig to 600s.
Removed specific Kconfig values from prj*.conf files.

oleo65 committed:
Renamed basic cluster attributes to comply with library defaults.

oleo65 committed:
Added hardware version config value.

oleo65 committed:
Slightly improved power cluster with more values.

oleo65 commented:
I pushed changes with the discussed config namings and some slight improvements to the power cluster. This should resolve the discussions.

I did not change the `CONFIG_LOG_DEFAULT_LEVEL` as this is a default zephyr config flag I reused in the `main.c` which should be fine for the main entry point.

rbaron commented:
LGTM. Let me know if you have any thoughts on the conversation I just unresolved:
- Board definitions doesn't seem like a good place for a `LOG` config for `prstlib`
- I'm not sure what `2` is in the default. How does it relate to `LOG_{DBG,INF,WRN,ERR}`? What should I set this to if I want to see `LOG_DBG` calls? If no great solution, please leave a comment with the mapping

Also okay to merge as is and I'll try some ideas. Tks!

oleo65 commented:
For the latter: 2 equals WRN. I found the values in the docstring of the zephyr library and I will add a comment to the config value and a reference where to find more. 😉 

The problem with the location of the prstlib log flag is tricky. If you put it within the sample Kconfig and reference it in the common prstlib, you need to "duplicate" the value in all samples (ble, blinky, etc.) otherwise they won't compile. We can definately do that but it feels also clunky to me.

I chose the board Kconfig, because that was the common place _within_ the prstlib folder. So I thought, that the config flag at least somewhat sits closer to the actual use in the prstlib. I don't know. 🤷

oleo65 committed:
Removed log level config from board definition.
Moved prst lib log level config to application layer with prst lib default fallback behaviour.

rbaron commented:
I think that's better, thanks. I will take a stab at setting up a proper `Kconfig` for prstlib. I think that would be the ideal solution. If the build system doesn't like it, I will at least set up a `prstlib/config.h` and move the config there, away from `macros.h`.

Tks once again!
